### ISSUE TITLE
Some question about nn_plist_dump

### ISSUE THREAD

**Issue by jwcesign** [eclipse-cyclonedds/cyclonedds#193]
Hi, Here I have some questions about plist dump, I use gdb to get this:
![image](https://user-images.githubusercontent.com/19745536/58797632-1842dd00-8633-11e9-9531-5f43982bd08b.png)

This problem makes some functions core dump for me, Can u check this, thanks a lot.

**Comment by eboasson** [comment]
Hi, would this be with 0644bb717f97915422c0548ef03e5d3289ef122b to make it possible to change the user data? And not by any chance with #192 applied? 🤔

It indeed looks like it should have copied it in ``nn_plist_dup`` and that ``new_plist`` contains garbage, but the value of the user_data field by itself is not enough to be sure: there is also a field named ``present`` that indicates which fields are set (and anything else is garbage) and a field named ``aliased`` that indicates whether the memory is owned by nn_plist_t/nn_xqos_t or not.

So, if this is reproducible, can you do me a favour and print ``datap.qos.present``, ``datap.qos.alised``, ``new_plist.qos.present`` and ``new_plist.qos.aliased`` in addition to ``datap.qos.user_data`` and ``new_plist.qos.user_data``? Note that ``present`` and ``aliased`` are a bitwise-or of the QP_xxx flags defined in q_xos.h. I would suspect datap.qos.present has ``QP_USER_DATA`` set.

Also, assuming this is with 0644bb717f97915422c0548ef03e5d3289ef122b in there, does this problem only happen when changing the user data?

**Comment by jwcesign** [comment]
Yes, I use 0644bb7, datap.qos.present, datap.qos.alised, new_plist.qos.present and new_plist.qos.aliased all zero.

**Comment by jwcesign** [comment]
It looks like the data is invalid, but not check whether it is valid.

**Comment by eboasson** [comment]
I've tried to get it to crash here, but I have so far failed in my attempts. Reading back, I realize that I am not 100% sure it crashes inside Cyclone for you. Maybe it is your application that crashes because it gets unexpected data out of ``dds_qget_userdata``: if the (proxy) participant's QoS object had default user data (an empty octet sequence), it is possible that the QoS object doesn't have the QP_USER_DATA flag set, and in that case ``dds_qget_userdata`` leaves its out parameters unchanged and returns false.

That that can happen when you look at the qos of a local or remote entity is a bug.

As a result of trying to understand your problem and seeing these things, I have now mostly implemented setting QoS values for any type of entity, though still with a few limitations. Hopefully I'll manage to remove those limitations this week (but no promises). Unfortunately, it seems I have confused CMake somehow/somewhere in a way that messed up building on Windows, but I think you're using Linux ... https://github.com/eboasson/cyclonedds/tree/set-qos

**Comment by eboasson** [comment]
@jwcesign I am hoping you've been able to work around the problem, it's been so quiet ... I just wanted to point out that that branch on my own fork that I mentioned has now resulted in PR #195. It would be interesting to me if you could see if it works well for you. Otherwise — unless anyone objects — I will assume the best and merge it in a day or two.

**Comment by jwcesign** [comment]
Hi, I am sorry for check so late, recently really busy. Yes, I want to use user_data to do some operation, if the length is wrong and value is invalid memory address. it will crash. Now I check it's because of this:
~~~c
#q_plist.c
void nn_plist_init_empty (nn_plist_t *dest)
{
#ifndef NDEBUG
  memset (dest, 0, sizeof (*dest));
#endif
  dest->present = dest->aliased = 0;
  nn_xqos_init_empty (&dest->qos);
}
~~~
As default, `NDEBUG` is defined, so `dest`, so the user_data length and value is invalid. Can u tell me why not set dest to zero?

**Comment by eboasson** [comment]
No worries about the delay — rather, thanks for letting me know.

Actually I think the memsetting to 0 is a mistake: that would hide bugs in a development build whereas the purpose would more likely be exposing them. So I think the idea was more to memset to something weird, like 0xdd, because that would be pretty likely to show up as an invalid value if it'd be used by accident anywhere.

Anyway, I have merged #195. Which means you no longer need that quick hack I did, and also that the QoS are all set in the QoS objects it returns via ``get_qos`` and the built-in topics. That probably wouldn't have happened just yet if you hadn't ran into this problem, so thank you for that ☺️
